File: common/src/main/java/com/pluxity/user/service/RoleService.java
package com.pluxity.user.service;

import com.pluxity.global.utils.SortUtils;
import com.pluxity.permission.PermissionGroup;
import com.pluxity.permission.PermissionGroupService;
import com.pluxity.permission.PermissionService;
import com.pluxity.user.dto.RoleCreateRequest;
import com.pluxity.user.dto.RoleResponse;
import com.pluxity.user.dto.RoleUpdateRequest;
import com.pluxity.user.entity.Role;
import com.pluxity.user.entity.RolePermission;
import com.pluxity.user.repository.RolePermissionRepository;
import com.pluxity.user.repository.RoleRepository;
import com.pluxity.user.repository.UserRoleRepository;
import jakarta.persistence.EntityManager;
import jakarta.persistence.EntityNotFoundException;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class RoleService {

    private final RoleRepository roleRepository;
    private final PermissionService permissionService;
    private final RolePermissionRepository rolePermissionRepository;
    private final UserRoleRepository userRoleRepository;
    private final PermissionGroupService permissionGroupService;
    private final EntityManager em;

    @Transactional
    public Long save(RoleCreateRequest request) {

        Role role = Role.builder().name(request.name()).description(request.description()).build();

        roleRepository.save(role);

        if (request.permissionGroupIds() != null && !request.permissionGroupIds().isEmpty()) {
            List<RolePermission> newRolePermissions =
                    request.permissionGroupIds().stream()
                            .map(
                                    permissionGroupId -> {
                                        PermissionGroup permissionGroup =
                                                permissionGroupService.findPermissionGroupById(permissionGroupId);
                                        return RolePermission.builder()
                                                .permissionGroup(permissionGroup)
                                                .role(role)
                                                .build();
                                    })
                            .toList();

            rolePermissionRepository.saveAll(newRolePermissions);
            newRolePermissions.forEach(role::addRolePermission);
        }

        return role.getId();
    }

    @Transactional(readOnly = true)
    public RoleResponse findById(Long id) {
        return RoleResponse.from(findRoleById(id));
    }

    @Transactional(readOnly = true)
    public List<RoleResponse> findAll() {
        return roleRepository.findAll(SortUtils.getOrderByCreatedAtDesc()).stream()
                .map(RoleResponse::from)
                .toList();
    }

    @Transactional
    public void update(Long id, RoleUpdateRequest request) {
        Role role = findRoleById(id);

        if (request.name() != null && !request.name().isBlank()) {
            role.changeRoleName(request.name());
        }
        if (request.description() != null) {
            role.changeDescription(request.description());
        }

        if (request.permissionGroupIds() != null) {
            syncPermissionGroups(role, request.permissionGroupIds());
        }
    }

    private void syncPermissionGroups(Role role, List<Long> requestedGroupIds) {
        Set<Long> currentGroupIds =
                role.getRolePermissions().stream()
                        .map(rp -> rp.getPermissionGroup().getId())
                        .collect(Collectors.toSet());
        Set<Long> requestedGroupIdsSet = new HashSet<>(requestedGroupIds);

        List<RolePermission> rolePermissionsToRemove =
                role.getRolePermissions().stream()
                        .filter(rp -> !requestedGroupIdsSet.contains(rp.getPermissionGroup().getId()))
                        .toList();

        if (!rolePermissionsToRemove.isEmpty()) {
            rolePermissionRepository.deleteAllInBatch(rolePermissionsToRemove);
            rolePermissionsToRemove.forEach(role::removeRolePermission);
        }

        Set<Long> idsToAdd =
                requestedGroupIdsSet.stream()
                        .filter(id -> !currentGroupIds.contains(id))
                        .collect(Collectors.toSet());

        if (!idsToAdd.isEmpty()) {
            List<RolePermission> rolePermissionsToAdd =
                    idsToAdd.stream()
                            .map(
                                    groupId -> {
                                        PermissionGroup permissionGroup =
                                                permissionGroupService.findPermissionGroupById(groupId);
                                        return RolePermission.builder()
                                                .role(role)
                                                .permissionGroup(permissionGroup)
                                                .build();
                                    })
                            .toList();

            List<RolePermission> savedRolePermissions =
                    rolePermissionRepository.saveAll(rolePermissionsToAdd);
            savedRolePermissions.forEach(role::addRolePermission);
        }
    }

    @Transactional
    public void delete(Long id) {
        Role role = findRoleById(id);
        rolePermissionRepository.deleteAllByRole(role);
        userRoleRepository.deleteAllByRole(role);
        em.flush();
        em.clear();
        roleRepository.deleteById(role.getId());
    }

    public Role findRoleById(Long id) {
        return roleRepository
                .findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Role not found with id: " + id));
    }
}

File: common/src/main/java/com/pluxity/user/service/RoleService.kt
package com.pluxity.user.service

import com.pluxity.permission.PermissionGroupService
import com.pluxity.permission.PermissionService
import com.pluxity.user.dto.RoleCreateRequest
import com.pluxity.user.dto.RoleResponse
import com.pluxity.user.dto.RoleUpdateRequest
import com.pluxity.user.dto.toRoleResponse
import com.pluxity.user.entity.Role
import com.pluxity.user.entity.RolePermission
import com.pluxity.user.repository.RolePermissionRepository
import com.pluxity.user.repository.RoleRepository
import com.pluxity.user.repository.UserRoleRepository
import jakarta.persistence.EntityManager
import jakarta.persistence.EntityNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class RoleService(
    private val roleRepository: RoleRepository,
    private val permissionService: PermissionService,
    private val rolePermissionRepository: RolePermissionRepository,
    private val userRoleRepository: UserRoleRepository,
    private val permissionGroupService: PermissionGroupService,
    private val em: EntityManager,
) {
    @Transactional
    fun save(request: RoleCreateRequest): Long {
        val role =
            roleRepository.save(
                Role(
                    name = request.name,
                    description = request.description,
                ),
            )

        request.permissionGroupIds.let { groupIds ->
            if (groupIds.isNotEmpty()) {
                val newRolePermissions =
                    groupIds.map { groupId ->
                        val permissionGroup = permissionGroupService.findPermissionGroupById(groupId)
                        RolePermission(
                            role = role,
                            permissionGroup = permissionGroup,
                        )
                    }

                rolePermissionRepository.saveAll(newRolePermissions)
                newRolePermissions.forEach { rolePermission ->
                    role.addRolePermission(rolePermission)
                }
            }
        }

        return role.id!!
    }

    @Transactional(readOnly = true)
    fun findById(id: Long): RoleResponse = findRoleById(id).toRoleResponse()

    @Transactional(readOnly = true)
    fun findAll(): List<RoleResponse> =
        roleRepository
            .findByAuthIsNotOrderByCreatedAtDesc("ADMIN")
            .map { it.toRoleResponse() }

    @Transactional
    fun update(
        id: Long,
        request: RoleUpdateRequest,
    ) {
        val role = findRoleById(id)

        request.name?.takeIf { it.isNotBlank() }?.let {
            role.changeRoleName(it)
        }
        request.description?.let { role.changeDescription(request.description) }

        request.permissionGroupIds?.let { syncPermissionGroups(role, request.permissionGroupIds) }
    }

    private fun syncPermissionGroups(
        role: Role,
        requestedGroupIds: List<Long>,
    ) {
        val currentGroupIds = role.rolePermissions.map { it.permissionGroup.id }.toSet()
        val requestedGroupIdsSet = requestedGroupIds.toSet()

        val rolePermissionsToRemove =
            role.rolePermissions
                .filter { !requestedGroupIdsSet.contains(it.permissionGroup.id) }

        if (rolePermissionsToRemove.isNotEmpty()) {
            rolePermissionRepository.deleteAllInBatch(rolePermissionsToRemove)
            rolePermissionsToRemove.forEach { rolePermission ->
                role.removeRolePermission(rolePermission)
            }
        }

        val idsToAdd = requestedGroupIdsSet.filter { !currentGroupIds.contains(it) }

        if (idsToAdd.isNotEmpty()) {
            val rolePermissionsToAdd =
                idsToAdd.map { groupId ->
                    val permissionGroup = permissionGroupService.findPermissionGroupById(groupId)
                    RolePermission(
                        role = role,
                        permissionGroup = permissionGroup,
                    )
                }

            rolePermissionRepository.saveAll(rolePermissionsToAdd).forEach { rolePermission ->
                role.addRolePermission(rolePermission)
            }
        }
    }

    @Transactional
    fun delete(id: Long) {
        val role = findRoleById(id)
        rolePermissionRepository.deleteAllByRole(role)
        userRoleRepository.deleteAllByRole(role)
        em.flush()
        em.clear()
        roleRepository.deleteById(role.id!!)
    }

    fun findRoleById(id: Long): Role =
        roleRepository.findWithInfoById(id)
            ?: throw EntityNotFoundException("Role not found with id: $id")
}
